### PR TITLE
[CC-6218] users list from local/remote xml file

### PR DESCRIPTION
**Pull request by jacek-klapacz**
__Story / Bug id:__ https://bugzilla.stxnext.pl/show_bug.cgi?id=6218
__Project / App:__ wdrozenie
__Reference person:__ @JacekKlapacz
__Description:__

read users list from local xmlfile, update local xml file from remote server by cron script
cron script: ./bin/cron_jobs_update_xml

__Results:__
pep8: empty

Report
======
225 statements analysed.

Statistics by type
------------------

+---------+-------+-----------+-----------+------------+---------+
|type     |number |old number |difference |%documented |%badname |
+=========+=======+===========+===========+============+=========+
|module   |7      |7          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|class    |2      |2          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|method   |18     |18         |=          |100.00      |50.00    |
+---------+-------+-----------+-----------+------------+---------+
|function |19     |19         |=          |100.00      |5.26     |
+---------+-------+-----------+-----------+------------+---------+

External dependencies
---------------------
::

    flask (presence_analyzer.main,presence_analyzer.views,presence_analyzer.utils)
    lxml 
      \-etree (presence_analyzer.utils)
    presence_analyzer 
      \-main (presence_analyzer.cron_jobs,presence_analyzer.tests,presence_analyzer.views,presence_analyzer,presence_analyzer.utils)
      \-utils (presence_analyzer.tests,presence_analyzer.views)
      \-views (presence_analyzer)

Raw metrics
-----------

+----------+-------+------+---------+-----------+
|type      |number |%     |previous |difference |
+==========+=======+======+=========+===========+
|code      |283    |53.70 |283      |=          |
+----------+-------+------+---------+-----------+
|docstring |159    |30.17 |159      |=          |
+----------+-------+------+---------+-----------+
|comment   |8      |1.52  |8        |=          |
+----------+-------+------+---------+-----------+
|empty     |77     |14.61 |77       |=          |
+----------+-------+------+---------+-----------+

Duplication
-----------

+-------------------------+------+---------+-----------+
|                         |now   |previous |difference |
+=========================+======+=========+===========+
|nb duplicated lines      |0     |0        |=          |
+-------------------------+------+---------+-----------+
|percent duplicated lines |0.000 |0.000    |=          |
+-------------------------+------+---------+-----------+

Messages by category
--------------------

+-----------+-------+---------+-----------+
|type       |number |previous |difference |
+===========+=======+=========+===========+
|convention |10     |10       |=          |
+-----------+-------+---------+-----------+
|refactor   |0      |0        |=          |
+-----------+-------+---------+-----------+
|warning    |1      |1        |=          |
+-----------+-------+---------+-----------+
|error      |1      |1        |=          |
+-----------+-------+---------+-----------+

% errors / warnings by module
-----------------------------

+------------------------+-------+--------+---------+-----------+
|module                  |error  |warning |refactor |convention |
+========================+=======+========+=========+===========+
|presence_analyzer.utils |100.00 |0.00    |0.00     |0.00       |
+------------------------+-------+--------+---------+-----------+
|presence_analyzer.main  |0.00   |100.00  |0.00     |0.00       |
+------------------------+-------+--------+---------+-----------+
|presence_analyzer.tests |0.00   |0.00    |0.00     |90.00      |
+------------------------+-------+--------+---------+-----------+
|presence_analyzer.views |0.00   |0.00    |0.00     |10.00      |
+------------------------+-------+--------+---------+-----------+

Messages
--------

+-----------------+------------+
|message id       |occurrences |
+=================+============+
|invalid-name     |10          |
+-----------------+------------+
|locally-disabled |5           |
+-----------------+------------+
|unused-import    |1           |
+-----------------+------------+
|no-member        |1           |
+-----------------+------------+
|file-ignored     |1           |
+-----------------+------------+

Global evaluation
-----------------
Your code has been rated at 9.29/10 (previous run: 9.29/10, +0.00)

__Migrations:__
N/A

__Rollback:__
N/A

__New imports / dependencies:__
N/A

__Unit Tests:__
Global evaluation
-----------------
Your code has been rated at 9.29/10 (previous run: 9.29/10, +0.00)

jklapacz@stxnext-010:~/git-test/presence-analyzer-jklapacz$ ./bin/test
Test parsing of CSV file. ... ok
Test mean presence time of given user. ... ok
Testing wrong user id: (string). ... ok
Testing wrong user id: (int). ... ok
Test mean start & end time of given user grouped by weekday ... ok
Testing wrong user id: (string). ... ok
Testing wrong user id: (int). ... ok
Test presence time of given user. ... ok
Testing wrong user id. ... ok
Test users listing. ... ok
Test main page redirect. ... ok
Test template: mean_time_weekday.html ... ok
Test template: presence_start_end.html ... ok
Test template: presence_weekday.html ... ok

----------------------------------------------------------------------
Ran 14 tests in 0.085s

### CODE REVIEW

File: src/presence_analyzer/static/js/base.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ function parseInterval(value) {
         $.getJSON("/api/v1/users", function(result) {
             var dropdown = $("#user_id");
             $.each(result, function(item) {
-                dropdown.append($("<option />").val(this.user_id).text(this.name));
+                dropdown.append($("<option />").val(this.user_id).attr('avatar', this.avatar).text(this.name));
```
Comment (jacek-klapacz) [inline src/presence_analyzer/static/js/base.js]: some tricky way, but it works fine.

File: src/presence_analyzer/cron_jobs.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    ./cron_jobs_update_xml remote.server.com/path/to/users.xml
+    """
+    app.config.from_pyfile(
+        os.path.join(
+            os.path.dirname(__file__), '..', '..', 'parts', 'etc', 'deploy.cfg'
+        )
+    )
+
+    if len(sys.argv) == 2:
+        app.config.update({'REMOTE_USERS_XML_URL': sys.argv[1]})
```
Comment (tomaszwozniak) [inline src/presence_analyzer/cron_jobs.py]: app.config['REMOTE_USERS_XML_URL'] = sys ...

File: src/presence_analyzer/templates/presence_weekday.html [hunk 3]
```diff
@@ -4,13 +4,17 @@
...
-	    });
+	    	if (result.success) {
+		        var data = google.visualization.arrayToDataTable(result.data);
+		        var options = {};
+		        chart_div.show();
+		        loading.hide();
+		        var chart = new google.visualization.PieChart(chart_div[0]);
+		        chart.draw(data, options);
+	    	} else {
+	    		alert('No data for selected user!')	;
```
Comment (tomaszwozniak) [inline src/presence_analyzer/templates/presence_weekday.html]: extra spaces

File: src/presence_analyzer/tests.py [hunk 4]
```diff
@@ -82,16 +94,18 @@ def test_api_presence_weekday(self):
         self.assertEqual(resp.status_code, 200)
         self.assertEqual(resp.content_type, 'application/json')
         data = json.loads(resp.data)
-        self.assertEqual(len(data), 8)
-        self.assertEqual(data[0], ["Weekday", "Presence (s)"])
-        self.assertEqual(data[1], ["Mon", 0])
+        self.assertEqual(len(data), 2)
+        self.assertEqual(data['success'], True)
```
Comment (tomaszwozniak) [inline src/presence_analyzer/tests.py]: assertTrue

File: src/presence_analyzer/tests.py [hunk 5]
```diff
@@ -101,18 +115,26 @@ def test_api_mean_time_weekday(self):
         self.assertEqual(resp.status_code, 200)
         self.assertEqual(resp.content_type, 'application/json')
         data = json.loads(resp.data)
-        self.assertEqual(len(data), 7)
-        self.assertEqual(data[0], ["Mon", 0])
+        self.assertEqual(len(data), 2)
+        self.assertEqual(data['success'], True)
```
Comment (tomaszwozniak) [inline src/presence_analyzer/tests.py]: assertFalse

File: src/presence_analyzer/tests.py [hunk 6]
```diff
@@ -101,18 +115,26 @@ def test_api_mean_time_weekday(self):
...
 
+    def test_api_mean_time_weekday_wrong_user(self):
+        """
+        Testing wrong user id: (int).
+        """
+        resp = self.client.get('/api/v1/mean_time_weekday/100')
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp.content_type, 'application/json')
+        data = json.loads(resp.data)
+        self.assertEqual(data['success'], False)
```
Comment (tomaszwozniak) [inline src/presence_analyzer/tests.py]: ditto

File: src/presence_analyzer/tests.py [hunk 7]
```diff
@@ -121,15 +143,23 @@ def test_api_presence_start_end_weekday(self):
         self.assertEqual(resp.status_code, 200)
         self.assertEqual(resp.content_type, 'application/json')
         data = json.loads(resp.data)
-        self.assertEqual(len(data), 7)
-        self.assertEqual(data[0], ["Mon", 0, 0])
+        self.assertEqual(len(data), 2)
+        self.assertEqual(data['success'], True)
```
Comment (tomaszwozniak) [inline src/presence_analyzer/tests.py]: ditto

File: src/presence_analyzer/tests.py [hunk 8]
```diff
@@ -121,15 +143,23 @@ def test_api_presence_start_end_weekday(self):
...
         """
-        Testing wrong user id.
+        Testing wrong user id: (int).
         """
         resp = self.client.get('/api/v1/presence_start_end/100')
-        self.assertEqual(resp.status_code, 404)
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp.content_type, 'application/json')
+        data = json.loads(resp.data)
+        self.assertEqual(data['success'], False)
```
Comment (tomaszwozniak) [inline src/presence_analyzer/tests.py]: ditto

File: src/presence_analyzer/utils.py [hunk 9]
```diff
@@ -71,6 +71,26 @@ def get_data():
...
+def get_users():
+    """
+    Get users from xml local file
+    """
+    data = []
+    tree = etree.parse(app.config['USERS_XML'])
+    server = tree.find('server')
+    path = server.find('protocol').text + '://' + server.find('host').text
+
+    for user in tree.iter('user'):
```
Comment (tomaszwozniak) [inline src/presence_analyzer/utils.py]: try to do this by list comprehension

**Comment by tomaszwozniak** [comment]
again minor comments

**Comment by jacek-klapacz** [comment]
__Story / Bug id:__ https://bugzilla.stxnext.pl/show_bug.cgi?id=6218
__Project / App:__ wdrozenie
__Reference person:__ @JacekKlapacz
__Description:__

users list from local/remote xml file - code cosmetics

__Results:__
pep8: empty

Report
======
220 statements analysed.

Statistics by type
------------------

+---------+-------+-----------+-----------+------------+---------+
|type     |number |old number |difference |%documented |%badname |
+=========+=======+===========+===========+============+=========+
|module   |7      |7          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|class    |2      |2          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|method   |18     |18         |=          |100.00      |50.00    |
+---------+-------+-----------+-----------+------------+---------+
|function |19     |19         |=          |100.00      |5.26     |
+---------+-------+-----------+-----------+------------+---------+



External dependencies
---------------------
::

    flask (presence_analyzer.main,presence_analyzer.views,presence_analyzer.utils)
    lxml 
      \-etree (presence_analyzer.utils)
    presence_analyzer 
      \-main (presence_analyzer.cron_jobs,presence_analyzer.tests,presence_analyzer.views,presence_analyzer,presence_analyzer.utils)
      \-utils (presence_analyzer.tests,presence_analyzer.views)
      \-views (presence_analyzer)



Raw metrics
-----------

+----------+-------+------+---------+-----------+
|type      |number |%     |previous |difference |
+==========+=======+======+=========+===========+
|code      |280    |53.13 |281      |-1.00      |
+----------+-------+------+---------+-----------+
|docstring |162    |30.74 |162      |=          |
+----------+-------+------+---------+-----------+
|comment   |8      |1.52  |8        |=          |
+----------+-------+------+---------+-----------+
|empty     |77     |14.61 |77       |=          |
+----------+-------+------+---------+-----------+



Duplication
-----------

+-------------------------+------+---------+-----------+
|                         |now   |previous |difference |
+=========================+======+=========+===========+
|nb duplicated lines      |0     |0        |=          |
+-------------------------+------+---------+-----------+
|percent duplicated lines |0.000 |0.000    |=          |
+-------------------------+------+---------+-----------+



Messages by category
--------------------

+-----------+-------+---------+-----------+
|type       |number |previous |difference |
+===========+=======+=========+===========+
|convention |10     |10       |=          |
+-----------+-------+---------+-----------+
|refactor   |0      |0        |=          |
+-----------+-------+---------+-----------+
|warning    |1      |1        |=          |
+-----------+-------+---------+-----------+
|error      |1      |1        |=          |
+-----------+-------+---------+-----------+



% errors / warnings by module
-----------------------------

+------------------------+-------+--------+---------+-----------+
|module                  |error  |warning |refactor |convention |
+========================+=======+========+=========+===========+
|presence_analyzer.utils |100.00 |0.00    |0.00     |0.00       |
+------------------------+-------+--------+---------+-----------+
|presence_analyzer.main  |0.00   |100.00  |0.00     |0.00       |
+------------------------+-------+--------+---------+-----------+
|presence_analyzer.tests |0.00   |0.00    |0.00     |90.00      |
+------------------------+-------+--------+---------+-----------+
|presence_analyzer.views |0.00   |0.00    |0.00     |10.00      |
+------------------------+-------+--------+---------+-----------+



Messages
--------

+-----------------+------------+
|message id       |occurrences |
+=================+============+
|invalid-name     |10          |
+-----------------+------------+
|locally-disabled |5           |
+-----------------+------------+
|unused-import    |1           |
+-----------------+------------+
|no-member        |1           |
+-----------------+------------+
|file-ignored     |1           |
+-----------------+------------+



Global evaluation
-----------------
Your code has been rated at 9.27/10 (previous run: 9.28/10, -0.00)


__Migrations:__
N/A

__Rollback:__
N/A

__New imports / dependencies:__
N/A

__Unit Tests:__
Test parsing of CSV file. ... ok
Test mean presence time of given user. ... ok
Testing wrong user id: (string). ... ok
Testing wrong user id: (int). ... ok
Test mean start & end time of given user grouped by weekday ... ok
Testing wrong user id: (string). ... ok
Testing wrong user id: (int). ... ok
Test presence time of given user. ... ok
Testing wrong user id. ... ok
Test users listing. ... ok
Test main page redirect. ... ok
Test template: mean_time_weekday.html ... ok
Test template: presence_start_end.html ... ok
Test template: presence_weekday.html ... ok

----------------------------------------------------------------------
Ran 14 tests in 0.084s

File: src/presence_analyzer/utils.py [hunk 10]
```diff
@@ -71,6 +71,26 @@ def get_data():
...
+    """
+    tree = etree.parse(app.config['USERS_XML'])
+    server = tree.find('server')
+    path = server.find('protocol').text + '://' + server.find('host').text
+
+    data = [
+        {
+            'name': user.find('name').text,
+            'user_id': user.attrib['id'],
+            'avatar': path + user.find('avatar').text
```
Comment (tomaszwozniak) [inline src/presence_analyzer/utils.py]: usually we use format but in this case we can leave it

File: src/presence_analyzer/utils.py [hunk 11]
```diff
@@ -71,6 +71,26 @@ def get_data():
...
 
+def get_users():
+    """
+    Get users from xml local file
+    """
+    tree = etree.parse(app.config['USERS_XML'])
+    server = tree.find('server')
+    path = server.find('protocol').text + '://' + server.find('host').text
+
+    data = [
```
Comment (tomaszwozniak) [inline src/presence_analyzer/utils.py]: return [

File: src/presence_analyzer/utils.py [hunk 12]
```diff
@@ -71,6 +71,26 @@ def get_data():
     return data
 
 
+def get_users():
+    """
+    Get users from xml local file
+    """
+    tree = etree.parse(app.config['USERS_XML'])
+    server = tree.find('server')
+    path = server.find('protocol').text + '://' + server.find('host').text
```
Comment (tomaszwozniak) [inline src/presence_analyzer/utils.py]: please use format

File: src/presence_analyzer/views.py [hunk 13]
```diff
@@ -53,11 +53,8 @@ def users_json_view():
     """
     Users listing for dropdown.
     """
-    data = utils.get_data()
-    return [
-        {'user_id': i, 'name': 'User {0}'.format(str(i))}
-        for i in data.keys()
-    ]
+    data = utils.get_users()
```
Comment (tomaszwozniak) [inline src/presence_analyzer/views.py]: return utils.get_users()

**Comment by tomaszwozniak** [comment]
Review finished

File: src/presence_analyzer/utils.py (hunk 11 above)
Comment (jacek-klapacz) [inline src/presence_analyzer/utils.py]: return [] it's faster/use less memory, or its only coding manner?

Comment (tomaszwozniak) [inline src/presence_analyzer/utils.py]: more second one

**Comment by jacek-klapacz** [comment]
__Story / Bug id:__ https://bugzilla.stxnext.pl/show_bug.cgi?id=6218
__Project / App:__ wdrozenie
__Reference person:__ @JacekKlapacz
__Description:__

users list from local/remote xml file - code cosmetics

__Results:__
pep8: empty

Report
======
217 statements analysed.

Statistics by type
------------------

+---------+-------+-----------+-----------+------------+---------+
|type     |number |old number |difference |%documented |%badname |
+=========+=======+===========+===========+============+=========+
|module   |7      |7          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|class    |2      |2          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|method   |18     |18         |=          |100.00      |50.00    |
+---------+-------+-----------+-----------+------------+---------+
|function |19     |19         |=          |100.00      |5.26     |
+---------+-------+-----------+-----------+------------+---------+



External dependencies
---------------------
::

    flask (presence_analyzer.main,presence_analyzer.views,presence_analyzer.utils)
    lxml 
      \-etree (presence_analyzer.utils)
    presence_analyzer 
      \-main (presence_analyzer.cron_jobs,presence_analyzer.tests,presence_analyzer.views,presence_analyzer,presence_analyzer.utils)
      \-utils (presence_analyzer.tests,presence_analyzer.views)
      \-views (presence_analyzer)



Raw metrics
-----------

+----------+-------+------+---------+-----------+
|type      |number |%     |previous |difference |
+==========+=======+======+=========+===========+
|code      |280    |53.13 |280      |=          |
+----------+-------+------+---------+-----------+
|docstring |162    |30.74 |162      |=          |
+----------+-------+------+---------+-----------+
|comment   |8      |1.52  |8        |=          |
+----------+-------+------+---------+-----------+
|empty     |77     |14.61 |77       |=          |
+----------+-------+------+---------+-----------+



Duplication
-----------

+-------------------------+------+---------+-----------+
|                         |now   |previous |difference |
+=========================+======+=========+===========+
|nb duplicated lines      |0     |0        |=          |
+-------------------------+------+---------+-----------+
|percent duplicated lines |0.000 |0.000    |=          |
+-------------------------+------+---------+-----------+



Messages by category
--------------------

+-----------+-------+---------+-----------+
|type       |number |previous |difference |
+===========+=======+=========+===========+
|convention |10     |10       |=          |
+-----------+-------+---------+-----------+
|refactor   |0      |0        |=          |
+-----------+-------+---------+-----------+
|warning    |0      |0        |=          |
+-----------+-------+---------+-----------+
|error      |1      |1        |=          |
+-----------+-------+---------+-----------+



% errors / warnings by module
-----------------------------

+------------------------+-------+--------+---------+-----------+
|module                  |error  |warning |refactor |convention |
+========================+=======+========+=========+===========+
|presence_analyzer.utils |100.00 |0.00    |0.00     |0.00       |
+------------------------+-------+--------+---------+-----------+
|presence_analyzer.tests |0.00   |0.00    |0.00     |90.00      |
+------------------------+-------+--------+---------+-----------+
|presence_analyzer.views |0.00   |0.00    |0.00     |10.00      |
+------------------------+-------+--------+---------+-----------+



Messages
--------

+-----------------+------------+
|message id       |occurrences |
+=================+============+
|invalid-name     |10          |
+-----------------+------------+
|locally-disabled |5           |
+-----------------+------------+
|no-member        |1           |
+-----------------+------------+
|file-ignored     |1           |
+-----------------+------------+



Global evaluation
-----------------
Your code has been rated at 9.31/10 (previous run: 9.31/10, +0.00)


__Migrations:__
N/A

__Rollback:__
N/A

__New imports / dependencies:__
N/A

__Unit Tests:__
Test parsing of CSV file. ... ok
Test mean presence time of given user. ... ok
Testing wrong user id: (string). ... ok
Testing wrong user id: (int). ... ok
Test mean start & end time of given user grouped by weekday ... ok
Testing wrong user id: (string). ... ok
Testing wrong user id: (int). ... ok
Test presence time of given user. ... ok
Testing wrong user id. ... ok
Test users listing. ... ok
Test main page redirect. ... ok
Test template: mean_time_weekday.html ... ok
Test template: presence_start_end.html ... ok
Test template: presence_weekday.html ... ok

----------------------------------------------------------------------
Ran 14 tests in 0.085s

OK

**Pull request merged**